lib/ClangImporter/ClangIncludePaths.cpp

```diff
@@ -296,14 +297,117 @@ getLibStdCxxFileMapping(ASTContext &ctx) {
+    llvm::SmallString<261> path{SDKPath};
+    llvm::sys::path::append(path, "usr", "share", modulemap);
+    if (llvm::sys::fs::exists(path))
+      return path.str().str();
+  }
+
+  if (!Options.RuntimeResourcePath.empty()) {
+    llvm::SmallString<261> path{Options.RuntimeResourcePath};
+    llvm::sys::path::append(path, "windows", modulemap);
+    if (llvm::sys::fs::exists(path))
+      return path.str().str();
+  }
+
+  return "";
+}
+
+SmallVector<std::pair<std::string, std::string>, 2>
+GetWindowsFileMappings(ASTContext &Context) {
+  const llvm::Triple &Triple = Context.LangOpts.Target;
+  const SearchPathOptions &SearchPathOpts = Context.SearchPathOpts;
+  SmallVector<std::pair<std::string, std::string>, 2> Mappings;
+  std::string AuxiliaryFile;
+
+  if (!Triple.isWindowsMSVCEnvironment())
+    return Mappings;
+
+  clang::driver::Driver Driver = createClangDriver(Context);
+  const llvm::opt::InputArgList Args = createClangArgs(Context, Driver);
+  const clang::driver::ToolChain &ToolChain = Driver.getToolChain(Args, Triple);
+  llvm::vfs::FileSystem &VFS = ToolChain.getVFS();
+
+  struct {
+    std::string Path;
+    std::string IncludeVersion;
+    std::string LibraryVersion;
+    int MajorVersion;
+  } WindowsSDK;
+  if (llvm::getWindowsSDKDir(VFS, {}, {}, {},
+                             WindowsSDK.Path, WindowsSDK.MajorVersion,
+                             WindowsSDK.IncludeVersion,
+                             WindowsSDK.LibraryVersion)) {
+    llvm::SmallString<261> WinSDKInjection{WindowsSDK.Path};
+    llvm::sys::path::append(WinSDKInjection, "Include");
+    if (WindowsSDK.MajorVersion > 8)
+      llvm::sys::path::append(WinSDKInjection, WindowsSDK.IncludeVersion, "um");
+    llvm::sys::path::append(WinSDKInjection, "module.modulemap");
+
+    AuxiliaryFile = GetWindowsAuxiliaryFile("winsdk.modulemap", SearchPathOpts);
+    if (!AuxiliaryFile.empty())
+      Mappings.emplace_back(std::string(WinSDKInjection), AuxiliaryFile);
+  }
+
+  struct {
+    std::string Path;
+    std::string Version;
+  } UCRTSDK;
+  if (llvm::getUniversalCRTSdkDir(VFS, {}, {}, {},
+                                  UCRTSDK.Path, UCRTSDK.Version)) {
+    llvm::SmallString<261> UCRTInjection{UCRTSDK.Path};
+    llvm::sys::path::append(UCRTInjection, "Include", UCRTSDK.Version, "ucrt");
+    llvm::sys::path::append(UCRTInjection, "module.modulemap");
+
+    AuxiliaryFile = GetWindowsAuxiliaryFile("ucrt.modulemap", SearchPathOpts);
+    if (!AuxiliaryFile.empty())
+      Mappings.emplace_back(std::string(UCRTInjection), AuxiliaryFile);
+  }
+
+  struct {
+    std::string Path;
+    llvm::ToolsetLayout Layout;
+  } VCTools;
+  if (llvm::findVCToolChainViaCommandLine(VFS, {}, {}, {}, VCTools.Path, VCTools.Layout) ||
+      llvm::findVCToolChainViaEnvironment(VFS, VCTools.Path, VCTools.Layout) ||
+      llvm::findVCToolChainViaSetupConfig(VFS, VCTools.Path, VCTools.Layout) ||
+      llvm::findVCToolChainViaRegistry(VCTools.Path, VCTools.Layout)) {
```

Comment:
I think this should be dropped. Modern Visual Studio (2017+) required by Swift doesn’t rely on registry. We may also want to assert `Layout` to be `VS2017OrNewer`.

Reply:
Huh, I missed that.  I like this idea, and yeah, we should be able to drop this check and add the assertion.